File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.cpp
class Solution {
public:
    //Linar search O(n)
    
    char nextGreatestLetter(vector<char>& letters, char target) {
        char ans = '\0';
        for(int i=0;i<letters.size();i++)
        {
            if(letters[i]>target)
            {
                ans = letters[i];
                break;
            }
        }
        if(ans!='\0')
        {
            return ans;
        }
        return letters[0];
    }
};

File: 0744-find-smallest-letter-greater-than-target/0744-find-smallest-letter-greater-than-target.java
class Solution {
    public int search(char[] letters,char x){
        int low = 0;
        int high = letters.length-1;
        int ind = 0;
        
        while(low<=high){
            int mid = low + (high-low)/2;
            if(letters[mid] <= x){
                low = mid+1;
            }
            else{
                ind = mid;
                high = mid-1;
            }
        }
        return ind;
    }
    
    public char nextGreatestLetter(char[] letters, char target) {
        int ans = search(letters,target);
        return letters[ans];
    }
}